bpf: nat: report detailed error from map_update_elem() for NAT entries

julianwiedmann opened the pull request:
Continuing from https://github.com/cilium/cilium/pull/24716, also report an `ext_err` when inserting entries into the NAT map fails.

```release-note
Report the kernel error code in case of packet drops due to failures to create NAT map entries.
```

julianwiedmann committed:
bpf: nat: report detailed error when creating SNAT entry

snat_*_nat_handle_mapping() already takes an *ext_err parameter to pass up
detailed error information from its CT logic. Use this to also report
errors from map_update_elem() when creating the NAT entries.

Signed-off-by: Julian Wiedmann <jwi@isovalent.com>

gentoo-root commented on bpf/lib/nat.h:
```diff
@@ -250,7 +250,7 @@ static __always_inline int snat_v4_new_mapping(struct __ctx_buff *ctx,
 					       const struct ipv4_nat_target *target,
 					       bool needs_ct)
 {
-	int ret = DROP_NAT_NO_MAPPING, retries;
```
I believe this change is wrong.

> If the very first lookup in the loop succeeds, we never update `ret` and
> it stays initialized to the initial value (!= 0).

If it succeeds, we set `ret = __snat_update(map, otuple, ostate, &rtuple, &rstate);` and break only on `ret == 0`, therefore we return 0, as we should.

On the other hand, if we exceed the number of retries, we should return DROP_NAT_NO_MAPPING, but with your change we would return 0.

julianwiedmann replied:
On success, `__snat_lookup()` returns the map entry. So `!__snat_lookup()` is false, and we don't step into the branch that touches `ret`.

gentoo-root replied:
Ah right, sorry, I got confused a bit. But I still don't get what the issue is that you are fixing.

If __snat_lookup returns non-NULL, on the first iteration, we change dport and try again. If it returns NULL on the second iteration, we set ret. If it doesn't return NULL on any iteration, we should return DROP_NAT_NO_MAPPING, not 0.

julianwiedmann replied:
> If __snat_lookup returns non-NULL, on the first iteration, we change dport and try again. If it returns NULL on the second iteration, we set ret. If it doesn't return NULL on any iteration, we should return DROP_NAT_NO_MAPPING, not 0.

Yep, agreed - this wouldn't fix anything. I must have convinced myself that we would bail out when finding an entry, but logic-wise that wouldn't make any sense either. So let's drop this part for now, I'll try to circle back during the 1.15 cycle and clean this up a bit. Thanks for keeping me honest, Maxim!

And sorry for not getting back to this earlier :(.

julianwiedmann committed:
bpf: dsr: report detailed error when creating RevDNAT entry

tail_nodeport_dsr_ingress_ipv*() already maintains an *ext_err parameter to
report detailed error information from its CT logic. Use this to also
report errors from map_update_elem() when creating the RevDNAT entry.

Signed-off-by: Julian Wiedmann <jwi@isovalent.com>

gentoo-root commented on bpf/lib/nat.h:
```diff
@@ -248,7 +248,14 @@ static __always_inline int snat_v4_new_mapping(struct __ctx_buff *ctx,
+
+	if (ret < 0) {
+		if (ext_err)
+			*ext_err = (__s8)ret;
```
ret could still be DROP_NAT_NO_MAPPING at this point, which doesn't make much sense as ext_err. On the other hand, ext_err < -128 is currently replaced by 0 in __DROP_REASON_EXT, so maybe it's not worth changing anything here.

julianwiedmann replied:
> ret could still be DROP_NAT_NO_MAPPING at this point, which doesn't make much sense as ext_err. On the other hand, ext_err < -128 is currently replaced by 0 in __DROP_REASON_EXT, so maybe it's not worth changing anything here.

Yep, it's a bit of an annoyance. But imho the `ext_err` always needs closer inspection/interpretation, based on where it comes from. So I think that's ok for now, and we can clean it up a bit more at a later time.

Thank you for the review, as always!
